livepatch: Ignore duplicate new symbols

For XenServer, we build cumulative live patches and replace so there is
only ever a single active one. This may cause duplicate new symbol
errors. Because of the lookup order (hypervisor first) and the fact that
live patches will only ever be built against the hypervisor (and not
another live patch), it should be safe to ignore these errors.

diff --git a/xen/arch/x86/Makefile b/xen/arch/x86/Makefile
index 9ab050a..9c92ce9 100644
--- a/xen/arch/x86/Makefile
+++ b/xen/arch/x86/Makefile
@@ -123,6 +123,12 @@ endif
 $(BASEDIR)/common/symbols-dummy.o:
 	$(MAKE) -f $(BASEDIR)/Rules.mk -C $(BASEDIR)/common symbols-dummy.o
 
+ifdef CONFIG_LIVEPATCH
+SYMBOLSFLAGS += --error-dup
+else
+SYMBOLSFLAGS += --warn-dup
+endif
+
 $(TARGET)-syms: prelink.o xen.lds $(BASEDIR)/common/symbols-dummy.o efi/relocs-dummy.o
 	$(LD) $(LDFLAGS) -T xen.lds -N prelink.o $(build_id_linker) efi/relocs-dummy.o \
 	    $(BASEDIR)/common/symbols-dummy.o -o $(@D)/.$(@F).0
@@ -133,7 +139,7 @@ $(TARGET)-syms: prelink.o xen.lds $(BASEDIR)/common/symbols-dummy.o efi/relocs-d
 	$(LD) $(LDFLAGS) -T xen.lds -N prelink.o $(build_id_linker) efi/relocs-dummy.o \
 	    $(@D)/.$(@F).0.o -o $(@D)/.$(@F).1
 	$(NM) -pa --format=sysv $(@D)/.$(@F).1 \
-		| $(BASEDIR)/tools/symbols $(all_symbols) --sysv --sort --warn-dup \
+		| $(BASEDIR)/tools/symbols $(all_symbols) --sysv --sort $(SYMBOLSFLAGS) \
 		>$(@D)/.$(@F).1.S
 	$(MAKE) -f $(BASEDIR)/Rules.mk $(@D)/.$(@F).1.o
 	$(LD) $(LDFLAGS) -T xen.lds -N prelink.o $(build_id_linker) efi/relocs-dummy.o \
diff --git a/xen/common/livepatch.c b/xen/common/livepatch.c
index 5da28a3..f5e490f 100644
--- a/xen/common/livepatch.c
+++ b/xen/common/livepatch.c
@@ -753,9 +753,6 @@ static int build_symbol_table(struct payload *payload,
             {
                 dprintk(XENLOG_ERR, LIVEPATCH "%s: duplicate new symbol: %s\n",
                         elf->name, symtab[i].name);
-                xfree(symtab);
-                xfree(strtab);
-                return -EEXIST;
             }
             symtab[i].new_symbol = 1;
             dprintk(XENLOG_DEBUG, LIVEPATCH "%s: new symbol %s\n",
diff --git a/xen/tools/symbols.c b/xen/tools/symbols.c
index 2fa6cb2..509d731 100644
--- a/xen/tools/symbols.c
+++ b/xen/tools/symbols.c
@@ -589,7 +589,7 @@ static int compare_name(const void *p1, const void *p2)
 int main(int argc, char **argv)
 {
 	unsigned int i;
-	bool unsorted = false, warn_dup = false, found_dup = false;
+	bool unsorted = false, warn_dup = false, error_dup = false, found_dup = false;
 
 	if (argc >= 2) {
 		for (i = 1; i < argc; i++) {
@@ -609,6 +609,8 @@ int main(int argc, char **argv)
 				sort_by_name = 1;
 			else if (strcmp(argv[i], "--warn-dup") == 0)
 				warn_dup = true;
+			else if (strcmp(argv[i], "--error-dup") == 0)
+				warn_dup = error_dup = true;
 			else
 				usage();
 		}
@@ -632,7 +634,7 @@ int main(int argc, char **argv)
 		unsorted = true;
 	}
 
-	if (found_dup)
+	if (error_dup && found_dup)
 		exit(1);
 
 	if (unsorted)
